src/System.Private.ServiceModel/src/Internals/System/Runtime/TimeoutHelper.cs

```diff
@@ -1,12 +1,69 @@
 // Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+using System.Collections.Concurrent;
 using System.Diagnostics.Contracts;
 using System;
 using System.Threading;
+using System.Threading.Tasks;
 
 namespace System.Runtime
 {
+    class CoalescedTimeoutTokens
+
+    {
+        private readonly uint _coalescingSpanMs;
+        private readonly ConcurrentDictionary<long, Task<CancellationToken>> _timeToToken =
+            new ConcurrentDictionary<long, Task<CancellationToken>>();
+
+        public CoalescedTimeoutTokens(int coalescingSpanMs = 15)
+        {
+            _coalescingSpanMs = (uint)coalescingSpanMs;
+        }
+
+        public CancellationToken FromTimeout(int millisecondsTimeout)
+        {
+            return FromTimeoutAsync(millisecondsTimeout).Result;
+        }
+
+        public Task<CancellationToken> FromTimeoutAsync(int millisecondsTimeout)
+        {
+            if (millisecondsTimeout <= 0)
+            {
+                return Task.FromResult(new CancellationToken(true));
+            }
+
+            uint currentTime = (uint)Environment.TickCount;
+            long targetTime = millisecondsTimeout + currentTime;
+            targetTime = ((targetTime + (_coalescingSpanMs - 1)) / _coalescingSpanMs) * _coalescingSpanMs;
+
+            Task<CancellationToken> tokenTask;
+
+            if (!_timeToToken.TryGetValue(targetTime, out tokenTask))
+            {
+                var tcs = new TaskCompletionSource<CancellationToken>();
+                if (_timeToToken.TryAdd(targetTime, tcs.Task))
+                {
+                    var token = new CancellationTokenSource((int)(targetTime - currentTime)).Token;
+                    token.Register(t => {
+                        Task<CancellationToken> ignored;
+                        _timeToToken.TryRemove((long)t, out ignored);
+                    }, targetTime);
+                    tcs.TrySetResult(token);
+                    tokenTask = tcs.Task;
```

Comment:
Why use a TaskCompletionSource? It's only ever set here synchronously, so why not just use Task.FromResult?

And then since everything here related to tasks is synchronous, I'd suggest removing tasks from it entirely.  If you really need to get a ```Task<CancellationToken>``` for some other future consumer, that consumer can then just use Task.FromResult with the CancellationToken this gives back.